src/cascadia/Remoting/WindowManager.cpp

```diff
@@ -77,57 +77,161 @@ namespace winrt::Microsoft::Terminal::Remoting::implementation
-                                  TraceLoggingPointer(nullptr, "Id", "No ID provided"),
-                                  TraceLoggingWideString(givenName.c_str(), "Name", "The name we should assign this window"),
+                                  "WindowManager_proposeToMonarch_unexpectedExceptionFromKing",
                                   TraceLoggingLevel(WINEVENT_LEVEL_VERBOSE),
                                   TraceLoggingKeyword(TIL_KEYWORD_TRACE));
+
+                LOG_CAUGHT_EXCEPTION();
+                throw;
             }
         }
+
+        // Here, the monarch (not us) has replied to the message. Get the
+        // valuables out of the response:
+        _shouldCreateWindow = result.ShouldCreateWindow();
+        if (result.Id())
+        {
+            givenID = result.Id().Value();
+        }
+        givenName = result.WindowName();
+
+        // TraceLogging doesn't have a good solution for logging an
+        // optional. So we have to repeat the calls here:
+        if (givenID)
+        {
+            TraceLoggingWrite(g_hRemotingProvider,
+                              "WindowManager_ProposeCommandline",
+                              TraceLoggingBoolean(_shouldCreateWindow, "CreateWindow", "true iff we should create a new window"),
+                              TraceLoggingUInt64(givenID.value(), "Id", "The ID we should assign our peasant"),
+                              TraceLoggingWideString(givenName.c_str(), "Name", "The name we should assign this window"),
+                              TraceLoggingLevel(WINEVENT_LEVEL_VERBOSE),
+                              TraceLoggingKeyword(TIL_KEYWORD_TRACE));
+        }
         else
+        {
+            TraceLoggingWrite(g_hRemotingProvider,
+                              "WindowManager_ProposeCommandline",
+                              TraceLoggingBoolean(_shouldCreateWindow, "CreateWindow", "true iff we should create a new window"),
+                              TraceLoggingPointer(nullptr, "Id", "No ID provided"),
+                              TraceLoggingWideString(givenName.c_str(), "Name", "The name we should assign this window"),
+                              TraceLoggingLevel(WINEVENT_LEVEL_VERBOSE),
+                              TraceLoggingKeyword(TIL_KEYWORD_TRACE));
+        }
+    }
+    void WindowManager::ProposeCommandline(const Remoting::CommandlineArgs& args)
+    {
+        // If we're the king, we _definitely_ want to process the arguments, we were
+        // launched with them!
+        //
+        // Otherwise, the King will tell us if we should make a new window
+        _shouldCreateWindow = _isKing;
+        std::optional<uint64_t> givenID;
+        winrt::hstring givenName{};
+        if (!_isKing)
+        {
+            _proposeToMonarch(args, givenID, givenName);
+        }
+
+        // During _proposeToMonarch, it's possible that we found that the king was dead, and we're the new king. Cool! Do this now.

```

Comment:
is there a chance that two peasants hit this at the exact same time and both think some moistened RPC call had lobbed a scimitar at them?

Reply:
I suppose this could be tested manually? But I still think that COM is only going to pick one of the two to be the new king